controllers/slicegateway/slicegateway.go

```diff
@@ -822,3 +889,102 @@ func contains(s []string, e string) bool {
 	}
 	return false
 }
+func getAddrSlice(nodeIPS []string) []corev1.EndpointAddress {
+	endpointSlice := make([]corev1.EndpointAddress, 0)
+	for _, ip := range nodeIPS {
+		endpointSlice = append(endpointSlice, corev1.EndpointAddress{IP: ip})
+	}
+	return endpointSlice
+}
+func validateEndpointAddresses(subset corev1.EndpointSubset, remoteNodeIPS []string) bool {
+	addrSlice := subset.Addresses
+	for _, address := range addrSlice {
+		if !contains(remoteNodeIPS, address.IP) {
+			return false
+		}
+	}
+	return true
+}
+
+// total -> external ip list of nodes in the k8s cluster
+// current -> ip list present in nodeIPs of cluster cr
+func validatenodeipcount(total, current []string) bool {
+	for _, a := range total {
+		if !contains(current, a) {
+			return false
+		}
+	}
+	return true && len(total) == len(current)
+}
+func UpdateGWPodStatus(gwPodStatus []*kubeslicev1beta1.GwPodInfo, podName string) []*kubeslicev1beta1.GwPodInfo {
+	index := -1
+	for i, _ := range gwPodStatus {
+		if gwPodStatus[i].PodName == podName {
+			index = i
+			break
+		}
+	}
+	return append(gwPodStatus[:index], gwPodStatus[index+1:]...)
+
+}
+func getLocalNSMIPs(slicegateway *kubeslicev1beta1.SliceGateway) []string {
+	nsmIPs := make([]string, 0)
+	for i, _ := range slicegateway.Status.GatewayPodStatus {
+		nsmIPs = append(nsmIPs, slicegateway.Status.GatewayPodStatus[i].LocalNsmIP)
+	}
+	return nsmIPs
+}
+func getLocalNSMIPsForRouter(slicegateway *kubeslicev1beta1.SliceGateway) []string {
+	nsmIPs := make([]string, 0)
+	for i, _ := range slicegateway.Status.GatewayPodStatus {
+		if slicegateway.Status.GatewayPodStatus[i].RouteRemoved == 1 {
+			continue
+		}
+		nsmIPs = append(nsmIPs, slicegateway.Status.GatewayPodStatus[i].LocalNsmIP)
+	}
+	return nsmIPs
+}
+func getPodIPs(slicegateway *kubeslicev1beta1.SliceGateway) []string {
+	podIPs := make([]string, 0)
+	for i, _ := range slicegateway.Status.GatewayPodStatus {
+		podIPs = append(podIPs, slicegateway.Status.GatewayPodStatus[i].PodIP)
+	}
+	return podIPs
+}
+func getPodNames(slicegateway *kubeslicev1beta1.SliceGateway) []string {
+	podNames := make([]string, 0)
+	for i, _ := range slicegateway.Status.GatewayPodStatus {
+		podNames = append(podNames, slicegateway.Status.GatewayPodStatus[i].PodName)
+	}
+	return podNames
+}
+func isGWPodStatusChanged(slicegateway *kubeslicev1beta1.SliceGateway, gwPod *kubeslicev1beta1.GwPodInfo) bool {
+	gwPodStatus := slicegateway.Status.GatewayPodStatus
+	for i, _ := range gwPodStatus {
+		if gwPodStatus[i].PodName == gwPod.PodName {
+			return gwPodStatus[i].TunnelStatus.Status == gwPod.TunnelStatus.Status
+		}
+	}
+	return false
+}
+func getPodAntiAffinity(slice string) *corev1.PodAntiAffinity {
+	return &corev1.PodAntiAffinity{
+		PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{{
+			Weight: 1,
+			PodAffinityTerm: corev1.PodAffinityTerm{
+				TopologyKey: controllers.TopologyKeySelector,
```

Comment:
The TopologyKey used here , does not exists on nodes
as per the docs:
"Note: Pod anti-affinity requires nodes to be consistently labelled, in other words, every node in the cluster must have an appropriate label matching topologyKey. If some or all nodes are missing the specified topologyKey label, it can lead to unintended behavior."

I suggest using:
http://kubernetes.io/hostname as the topologyKey

cc: @bharath-avesha wdyt?